deepcopy.go

```diff
@@ -103,41 +123,69 @@ func copyRecursive(original, cpy reflect.Value) {
 			if original.Type().Field(i).PkgPath != "" {
 				continue
 			}
-			copyRecursive(original.Field(i), cpy.Field(i))
+			copyRecursive(original.Field(i), cpy.Field(i), state)
 		}
 	
 	case reflect.Slice:
+		if state.ptrLevel > uint(circularLimitationCounter) {
```

Comment:
Why check cycle reference after this condition? It might be better to check cycle at the entrance of every encode call, so that we could report a cycle condition ASAP.

Reply:
The ways to get pointers of objects are different, for example, we could retrieve the `Map` and `Ptr` by `ptr := original.Interface()`, however, for the slice, both length and pointer need to be stored.
```go
			ptr := struct {
				ptr unsafe.Pointer
				len int
			}{unsafe.Pointer(original.Pointer()), original.Len()}
```
As a result, we put it inside the switch.

Reply:
I meat `if state.ptrLevel > uint(circularLimitationCounter) `, not the outer `switch case`.

Reply:
We check cycle reference after condition because the depth surges quickly once there is a circular reference. If we set up the condition here, we needn't to pay additional storages(ptr seen) for the normal cases.

Reply:
@Chronostasys I added the benchmark and it benefits some.

<details><summary>Benchmark Report</summary>

```
benchmark                                             old ns/op     new ns/op     delta
BenchmarkDeepCopy/deepcopy-12     3815          4433          +16.20%
BenchmarkDeepCopy/deepcopy-12     3802          4686          +23.25%
BenchmarkDeepCopy/deepcopy-12     3845          4482          +16.57%
BenchmarkDeepCopy/deepcopy-12     3973          5942          +49.56%
BenchmarkDeepCopy/deepcopy-12     3887          6031          +55.16%
BenchmarkDeepCopy/deepcopy-12     3938          5388          +36.82%
BenchmarkDeepCopy/deepcopy-12     4021          5605          +39.39%
BenchmarkDeepCopy/deepcopy-12     4460          5315          +19.17%
BenchmarkDeepCopy/deepcopy-12     4360          5344          +22.57%
BenchmarkDeepCopy/deepcopy-12     4157          5759          +38.54%
BenchmarkDeepCopy/deepcopy-12     4367          5568          +27.50%
BenchmarkDeepCopy/deepcopy-12     4257          5385          +26.50%
BenchmarkDeepCopy/deepcopy-12     4459          5303          +18.93%
BenchmarkDeepCopy/deepcopy-12     4413          5121          +16.04%
BenchmarkDeepCopy/deepcopy-12     4328          5143          +18.83%
BenchmarkDeepCopy/deepcopy-12     4396          5145          +17.04%
BenchmarkDeepCopy/deepcopy-12     4390          5455          +24.26%
BenchmarkDeepCopy/deepcopy-12     4405          5459          +23.93%
BenchmarkDeepCopy/deepcopy-12     4468          5317          +19.00%
BenchmarkDeepCopy/deepcopy-12     4372          5143          +17.63%

benchmark                                             old allocs     new allocs     delta
BenchmarkDeepCopy/deepcopy-12     185            190            +2.70%
BenchmarkDeepCopy/deepcopy-12     185            190            +2.70%
BenchmarkDeepCopy/deepcopy-12     185            190            +2.70%
BenchmarkDeepCopy/deepcopy-12     185            190            +2.70%
BenchmarkDeepCopy/deepcopy-12     185            190            +2.70%
BenchmarkDeepCopy/deepcopy-12     185            190            +2.70%
BenchmarkDeepCopy/deepcopy-12     185            190            +2.70%
BenchmarkDeepCopy/deepcopy-12     185            190            +2.70%
BenchmarkDeepCopy/deepcopy-12     185            190            +2.70%
BenchmarkDeepCopy/deepcopy-12     185            190            +2.70%
BenchmarkDeepCopy/deepcopy-12     185            190            +2.70%
BenchmarkDeepCopy/deepcopy-12     185            190            +2.70%
BenchmarkDeepCopy/deepcopy-12     185            190            +2.70%
BenchmarkDeepCopy/deepcopy-12     185            190            +2.70%
BenchmarkDeepCopy/deepcopy-12     185            190            +2.70%
BenchmarkDeepCopy/deepcopy-12     185            190            +2.70%
BenchmarkDeepCopy/deepcopy-12     185            190            +2.70%
BenchmarkDeepCopy/deepcopy-12     185            190            +2.70%
BenchmarkDeepCopy/deepcopy-12     185            190            +2.70%
BenchmarkDeepCopy/deepcopy-12     185            190            +2.70%

benchmark                                             old bytes     new bytes     delta
BenchmarkDeepCopy/deepcopy-12     4376          4456          +1.83%
BenchmarkDeepCopy/deepcopy-12     4376          4456          +1.83%
BenchmarkDeepCopy/deepcopy-12     4376          4456          +1.83%
BenchmarkDeepCopy/deepcopy-12     4376          4456          +1.83%
BenchmarkDeepCopy/deepcopy-12     4376          4456          +1.83%
BenchmarkDeepCopy/deepcopy-12     4376          4456          +1.83%
BenchmarkDeepCopy/deepcopy-12     4376          4456          +1.83%
BenchmarkDeepCopy/deepcopy-12     4376          4456          +1.83%
BenchmarkDeepCopy/deepcopy-12     4376          4456          +1.83%
BenchmarkDeepCopy/deepcopy-12     4376          4456          +1.83%
BenchmarkDeepCopy/deepcopy-12     4376          4456          +1.83%
BenchmarkDeepCopy/deepcopy-12     4376          4456          +1.83%
BenchmarkDeepCopy/deepcopy-12     4376          4456          +1.83%
BenchmarkDeepCopy/deepcopy-12     4376          4456          +1.83%
BenchmarkDeepCopy/deepcopy-12     4376          4456          +1.83%
BenchmarkDeepCopy/deepcopy-12     4376          4456          +1.83%
BenchmarkDeepCopy/deepcopy-12     4376          4456          +1.83%
BenchmarkDeepCopy/deepcopy-12     4376          4456          +1.83%
BenchmarkDeepCopy/deepcopy-12     4376          4456          +1.83%
BenchmarkDeepCopy/deepcopy-12     4376          4456          +1.83%

```
</details>

<details><summary>Benchmark Code</summary>

```go
package main

import (
	"os"
	"sync"
	"testing"
	"time"
	
	xdeepcopy "github.com/xieyuschen/deepcopy"
	"gopkg.in/yaml.v2"
)

func BenchmarkDeepCopy(b *testing.B) {
	var c AllTypeStructs
	yaml.Unmarshal([]byte(iniStr), &c)
	b.Run("deepcopy", func(b *testing.B) {
		b.ResetTimer()
		b.RunParallel(func(pb *testing.PB) {
			for pb.Next() {
				_, _ = xdeepcopy.Copy(&c)
			}
		})
	})
}

type AllTypeStructs struct {
	// unsupported, but wish to
	CircularRef *AllTypeStructs
	private     int
	Mutex       *sync.Mutex
	Chan        chan int `json:"-" yaml:"-"`
	Func        func()   `json:"-" yaml:"-"`
	File        *os.File
	Time        time.Time      `json:"time" yaml:"time"`
	Timep       *time.Time     `json:"timep" yaml:"timep"`
	Duration    time.Duration  `json:"duration" yaml:"duration"`
	Durationp   *time.Duration `json:"durationp" yaml:"durationp"`
	
	Int    int         `json:"int" yaml:"int"`
	Inti   interface{} `json:"inti" yaml:"inti"`
	Intp   *int        `json:"intp" yaml:"intp"`
	Slice  []int       `json:"slice" yaml:"slice"`
	SliceP *[]int      `json:"slicep" yaml:"slicep"`
	Array  [3]int      `json:"array" yaml:"array"`
	ArrayP *[3]int     `json:"arrayp" yaml:"arrayp"`
	
	Roster   map[int]HelperStruct   `json:"roster" yaml:"roster"`
	RosterP  map[int]*HelperStruct  `json:"rosterp" yaml:"rosterp"`
	PRoster  *map[int]HelperStruct  `json:"proster" yaml:"proster"`
	PRosterP *map[int]*HelperStruct `json:"prosterp" yaml:"prosterp"`
	
	Person *HelperStruct `json:"person" yaml:"person"`
	Pp     *Pointer      `json:"pp" yaml:"pp"`
	Ss     [1][][]int    `yaml:"ss"`
}

type Pointer struct {
	Intp        *int `json:"intp" yaml:"intp"`
	CircularRef *AllTypeStructs
}

type HelperStruct struct {
	Name string `json:"name" yaml:"name"`
	Age  *int   `json:"age" yaml:"age"`
}

const iniStr = `
ss: [[[1,2],[3,4]]]
time: 2018-01-09T10:40:47Z
timep: 2018-01-09T10:40:47Z
duration: 1h15m30.918273645s
durationp: 1h15m30.918273645s
int: 1
inti: 1
intp: 1
slice:
  - 1
  - 2
slicep:
  - 1
  - 2
array: [1,2,3]
arrayp: [1,2,3]
roster:
  1:
    name: tom
    age: 10
  2:
    name: jerry
    age: 10
rosterp:
  1:
    name: tom
    age: 10
  2:
    name: jerry
    age: 10
proster:
  1:
    name: tom
    age: 10
  2:
    name: jerry
    age: 10
prosterp:
  1:
    name: tom
    age: 10
  2:
    name: jerry
    age: 10
person:
  name: bob
  age: 12
pp:
  intp: 1
`
```

</details>